Add Blockbench case study (#5091)
GoogleChrome/developer.chrome.com: site/en/blog/how-blockbench-uses-the-eyedropper-api/index.md

---
layout: 'layouts/blog-post.njk'
title: 'How the 3D model editor Blockbench uses the EyeDropper API to let users choose colors from everywhere'
subhead: >
  The EyeDropper API provides a mechanism for creating an eyedropper tool that lets users sample colors from their screens, including outside of the browser window. Learn how the 3D model editor Blockbench uses this API in their codebase.
date: 2023-01-31
# updated: 2023-01-31
hero: image/8WbTDNrhLsU0El80frMBGE4eMCD3/6fIc5QdWayhjyXmIFEi6.png
alt: Computer created with Blockbench.
authors:
  - thomassteiner
  - jannispetersen
tags:
  - capabilities

>   - case-study

Author: ```suggestion
  - case-study
  - fugu-case-study
```

---

> [Blockbench](https://www.blockbench.net/) is a free, modern model editor for low-poly models with pixel art textures. If you have ever played [Minecraft](https://www.minecraft.net/), chances are you have seen assets that were created with Blockbench, for example the goat, which was added to Minecraft as part of the Caves and Cliffs Update, Part 1. The GPL version 3 licensed Blockbench code is [open source](https://github.com/JannisX11/blockbench) and maintained by a friendly community of developers.

Reviewer: I've been thinking about a way to rephrase this last sentence to make it describe Blockbench more accurately, since I am the only maintainer and the main developer, and the community is actively testing and providing feedback. Maybe the best solution is just to remove the last half sentence and turn it into "The GPL version 3 licensed Blockbench code is open source on GitHub."

Author: Will change accordingly.

{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/3NK0qVqDGnG7hTpaZpUS.png", alt="Minecraft sheep designed with Blockbench.", width="800", height="426" %}

> To try Blockbench, launch the app by navigating to [web.blockbench.net](https://web.blockbench.net/). As a PWA, you can install it to your desktop and launch it in a standalone window. If you are new to 3D modeling, you can start with one of the many [open source `.bbmodel` examples](https://github.com/search?q=path%3A*.bbmodel&type=code) you can find on GitHub (login required). For example, try the train model that you can see in the following screenshot.

Author: ```suggestion
To try Blockbench, launch the app by navigating to [web.blockbench.net](https://web.blockbench.net/). As a PWA, you can install it to your desktop and launch it in a standalone window. If you are new to 3D modeling, you can start with one of the many open source `.bbmodel` file examples that you can find using GitHub's code search. For example, try the train model that you can see in the following screenshot.
```

> To try Blockbench, launch the app by navigating to [web.blockbench.net](https://web.blockbench.net/). As a PWA, you can install it to your desktop and launch it in a standalone window. If you are new to 3D modeling, you can start with one of the many [open source `.bbmodel` examples](https://github.com/search?q=path%3A*.bbmodel&type=code) you can find on GitHub (login required). For example, try the train model that you can see in the following screenshot.

Reviewer: Hmm, I'm not sure why the GitHub search link works for you and doesn't work for me. On my end it just displays this message: 
![image](https://user-images.githubusercontent.com/29278436/215861641-f5eabf6e-5f4b-4cdb-a0d1-cd2c124ed5ec.png)

Author: OK, traced down to the new code search experiment. It's incompatible both ways, links to the old version don't work for the new version and vice versa. I will replace the link with text then. 

![image](https://user-images.githubusercontent.com/145676/216041704-447fbfc8-fedb-4217-961e-bdcb9c356ee1.png)

{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/hE98H7sZPBvB8iTwnLlm.png", alt="ALT_TEXT_HERE", width="800", height="557" %}

## The EyeDropper API

Among many other features, Blockbench offers a paint feature, so if you ever wanted to tag a train, now you can. Please don't do this in the real world. You can see that for my tagging, I used a bright orange.

{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/aECYxTYC8RZHglK31nFz.png", alt="ALT_TEXT_HERE", width="800", height="595" %}

This is actually an orange that I have extracted directly from the [macOS Ventura wallpaper](https://9to5mac.com/2022/10/05/macos-13-ventura-wallpaper-download-it-right-here/) through the [EyeDropper API](https://developer.mozilla.org/docs/Web/API/EyeDropper_API). As you can see in the following screenshot, the eye dropper (the orange circle left of the app window) can reach outside of the application straight into my desktop, or any other app that I may have open.

{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/YqL7rf4OUgnlROPugFuy.png", alt="ALT_TEXT_HERE", width="800", height="426" %}

> With Blockbench being open source, you can learn how the developers have implemented the API. The code in question is in [`JannisX11/blockbench/blob/master/js/texturing/color.js`](https://github.com/JannisX11/blockbench/blob/35ced3b3d094ffbf3f2b9548e82bc31c27e0cd05/js/texturing/color.js#L1034-L1049). Blockbench is also available as an Electron.js appou can see from the comment, it has special case handling for an [issue in Electron](https://github.com/electron/electron/issues/27980), where the color picker can't pick color outside of the window. With the web API, which you can see in the highlighted part of the code snippet, this is not an issue. Using the API is straightforward. Instantiate a new [`EyeDropper`](https://developer.mozilla.org/docs/Web/API/EyeDropper) instance and call its [`open()`](https://developer.mozilla.org/docs/Web/API/EyeDropper/open) method. This method resolves with an `sRGBHex` string representing the selected color, in [hexadecimal sRGB format](https://developer.mozilla.org/docs/Web/CSS/hex-color).

Reviewer: There seems to be a few words missing after "Blockbench is also available as an Electron.js app".

Author: Looks like a letter got swallowed. This is the fix: `Electron.js app. You`.

> With Blockbench being open source, you can learn how the developers have implemented the API. The code in question is in [`JannisX11/blockbench/blob/master/js/texturing/color.js`](https://github.com/JannisX11/blockbench/blob/35ced3b3d094ffbf3f2b9548e82bc31c27e0cd05/js/texturing/color.js#L1034-L1049). Blockbench is also available as an Electron.js appou can see from the comment, it has special case handling for an [issue in Electron](https://github.com/electron/electron/issues/27980), where the color picker can't pick color outside of the window. With the web API, which you can see in the highlighted part of the code snippet, this is not an issue. Using the API is straightforward. Instantiate a new [`EyeDropper`](https://developer.mozilla.org/docs/Web/API/EyeDropper) instance and call its [`open()`](https://developer.mozilla.org/docs/Web/API/EyeDropper/open) method. This method resolves with an `sRGBHex` string representing the selected color, in [hexadecimal sRGB format](https://developer.mozilla.org/docs/Web/CSS/hex-color).

Author: ```suggestion
With Blockbench being open source, you can learn how the developers have implemented the API. The code in question is in [`JannisX11/blockbench/blob/master/js/texturing/color.js`](https://github.com/JannisX11/blockbench/blob/35ced3b3d094ffbf3f2b9548e82bc31c27e0cd05/js/texturing/color.js#L1034-L1049). Blockbench is also available as an Electron.js app. You can see from the comment, it has special case handling for an [issue in Electron](https://github.com/electron/electron/issues/27980), where the color picker can't pick color outside of the window. With the web API, which you can see in the highlighted part of the code snippet, this is not an issue. Using the API is straightforward. Instantiate a new [`EyeDropper`](https://developer.mozilla.org/docs/Web/API/EyeDropper) instance and call its [`open()`](https://developer.mozilla.org/docs/Web/API/EyeDropper/open) method. This method resolves with an `sRGBHex` string representing the selected color, in [hexadecimal sRGB format](https://developer.mozilla.org/docs/Web/CSS/hex-color).
```

---
layout: 'layouts/blog-post.njk'
title: 'How the 3D model editor Blockbench uses the EyeDropper API to let users choose colors from everywhere'
subhead: >
  The EyeDropper API provides a mechanism for creating an eyedropper tool that lets users sample colors from their screens, including outside of the browser window. Learn how the 3D model editor Blockbench uses this API in their codebase.
date: 2023-01-31
# updated: 2023-01-31
hero: image/8WbTDNrhLsU0El80frMBGE4eMCD3/6fIc5QdWayhjyXmIFEi6.png
alt: Computer created with Blockbench.
authors:
  - thomassteiner
  - jannispetersen
tags:
  - capabilities
  - case-study
  - fugu-case-study
---

> [Blockbench](https://www.blockbench.net/) is a free, modern model editor for low-poly models with pixel art textures. If you have ever played [Minecraft](https://www.minecraft.net/), chances are you have seen assets that were created with Blockbench, for example the goat, which was added to Minecraft as part of the Caves and Cliffs Update, Part 1. The GPL version 3 licensed Blockbench code is [open source on GitHub](https://github.com/JannisX11/blockbench).

Author: ```suggestion

## Introduction

(This article is also available in form of a video.)

{% YouTube id="Y40vMQap9fs", startTime="394" %}

[Blockbench](https://www.blockbench.net/) is a free, modern model editor for low-poly models with pixel art textures. If you have ever played [Minecraft](https://www.minecraft.net/), chances are you have seen assets that were created with Blockbench, for example the goat, which was added to Minecraft as part of the Caves and Cliffs Update, Part 1. The GPL version 3 licensed Blockbench code is [open source on GitHub](https://github.com/JannisX11/blockbench).
```

---
layout: 'layouts/blog-post.njk'
title: 'How the 3D model editor Blockbench uses the EyeDropper API to let users choose colors from everywhere'
subhead: >
  The EyeDropper API provides a mechanism for creating an eyedropper tool that lets users sample colors from their screens, including outside of the browser window. Learn how the 3D model editor Blockbench uses this API in their codebase.

> date: 2023-01-31
> # updated: 2023-01-31

Author: ```suggestion
date: 2023-05-17
# updated: 2023-05-17
```
